File: game-client/Assets/Scripts/GameState/MockEngine/MatchEngine.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace QuizBattle.GameState.MockEngine
{
    /// Mirrors server/src/matchEngine/MatchEngine.ts — the single authoritative core the
    /// server also implements. This local copy exists to validate game feel on one
    /// device before networking is wired up (see project plan, Phase 1); once the Unity
    /// client talks to the real server, gameplay scripts should read state from
    /// MatchStateStore (server truth) instead of driving this engine directly.
    public static class MatchEngine
    {
        public static MatchState CreateMatch(int matchId, MatchMode mode, int maxRounds = 20, int? gridHeight = null) =>
            new MatchState(matchId, mode, maxRounds, gridHeight);

        public static PlayerState AddPlayer(MatchState state, int playerId, string name, string characterId, string team, GridPos startPos)
        {
            var character = CharacterCatalog.Get(characterId);
            var player = new PlayerState
            {
                playerId = playerId,
                name = name,
                characterId = characterId,
                team = team,
                hp = character.maxHp,
                maxHp = character.maxHp,
                pos = startPos,
                alive = true,
            };
            state.players[playerId] = player;
            return player;
        }

        public static void StartMatch(MatchState state)
        {
            state.status = MatchStatus.Active;
        }

        /// Moves a player toward the goal row (gridHeight - 1), clamped there, and flags
        /// goalReached once they hit it. The lane is a single dimension — x never changes
        /// after the player's starting position.
        private static void AdvancePlayer(PlayerState player, int gridHeight, int steps)
        {
            int goalY = gridHeight - 1;
            int nextY = Math.Min(goalY, player.pos.y + steps);
            player.pos = new GridPos(player.pos.x, nextY);
            if (nextY >= goalY) player.goalReached = true;
        }

        public static void PushQuestion(MatchState state, int playerId, int questionId, int correctIndex)
        {
            if (!state.players.TryGetValue(playerId, out var player)) return;
            player.currentQuestion = new ActiveQuestion { questionId = questionId, correctIndex = correctIndex, questionNumber = player.questionsAnswered + 1 };
        }

        public struct SubmitAnswerResult
        {
            public bool ok;
            public string error;
            public bool correct;
            public int streakCount;
            public PendingReward rewardOffered;
            public GridPos newPos;
            public bool goalReached;
            public int dotDamage;
            public MatchResult result;
        }

        public static SubmitAnswerResult SubmitAnswer(MatchState state, int playerId, int choiceIndex, Func<double> rng = null)
        {
            if (!state.players.TryGetValue(playerId, out var player) || !player.alive)
                return new SubmitAnswerResult { ok = false, error = "not_in_match" };
            if (player.currentQuestion == null)
                return new SubmitAnswerResult { ok = false, error = "no_active_question" };

            bool correct = choiceIndex == player.currentQuestion.correctIndex;
            player.currentQuestion = null;
            player.questionsAnswered += 1;

            player.consecutiveCorrect = correct ? player.consecutiveCorrect + 1 : 0;
            if (correct)
            {
                player.totalCorrectAnswers += 1;
                player.maxStreak = Math.Max(player.maxStreak, player.consecutiveCorrect);
            }

            PendingReward rewardOffered = null;
            if (correct && player.consecutiveCorrect >= 2 && player.pendingReward == null)
            {
                var reward = RewardResolver.RollReward(player, player.questionsAnswered, rng);
                player.pendingReward = reward;
                rewardOffered = reward;
            }
            RewardResolver.ExpireStaleRewards(new[] { player }, player.questionsAnswered);

            // A correct answer is what drives you toward the goal — this replaces the
            // old movement-budget/direction system entirely. A frozen player still
            // answers/builds streak normally; only this one advance is consumed/skipped.
            if (correct)
            {
                if (player.frozen)
                {
                    player.frozen = false;
                }
                else
                {
                    AdvancePlayer(player, state.gridHeight, 1);
                }
            }

            int dotDamage = CombatResolver.TickDot(player);

            var result = CheckWinCondition(state);
            state.result = result;
            if (result != null) state.status = MatchStatus.Completed;

            return new SubmitAnswerResult
            {
                ok = true,
                correct = correct,
                streakCount = player.consecutiveCorrect,
                rewardOffered = rewardOffered,
                newPos = player.pos,
                goalReached = player.goalReached,
                dotDamage = dotDamage,
                result = result,
            };
        }

        /// A player's personal question timer expired with no answer submitted — scored
        /// the same as a wrong answer (streak reset, no movement), but still ticks their
        /// own DoT and advances their personal question count.
        public static SubmitAnswerResult TimeoutAnswer(MatchState state, int playerId)
        {
            if (!state.players.TryGetValue(playerId, out var player) || !player.alive || player.currentQuestion == null)
                return new SubmitAnswerResult { ok = false, error = "no_active_question" };

            player.currentQuestion = null;
            player.questionsAnswered += 1;
            player.consecutiveCorrect = 0;

            // A reward the player never acted on (missed/ignored the popup) must not linger —
            // SubmitAnswer's pendingReward gate would otherwise silently block every future
            // streak reward until natural expiry, several questions later.
            if (player.pendingReward != null)
            {
                RewardResolver.ConsumeReward(player, player.pendingReward.rewardId);
            }
            RewardResolver.ExpireStaleRewards(new[] { player }, player.questionsAnswered);

            int dotDamage = CombatResolver.TickDot(player);

            var result = CheckWinCondition(state);
            state.result = result;
            if (result != null) state.status = MatchStatus.Completed;

            return new SubmitAnswerResult
            {
                ok = true,
                correct = false,
                streakCount = 0,
                rewardOffered = null,
                newPos = player.pos,
                goalReached = player.goalReached,
                dotDamage = dotDamage,
                result = result,
            };
        }

        /// Every living player attacker could legally target (alive, not self, not a
        /// teammate) other than excludeId — used both to enforce "can't target the same
        /// player twice in a row" and to waive that rule when there's genuinely no one
        /// else to pick (e.g. a 2-player match), rather than soft-locking the reward entirely.
        private static bool HasAlternativeTarget(MatchState state, PlayerState attacker, int excludeId)
        {
            foreach (var p in state.players.Values)
            {
                if (!p.alive || p.playerId == attacker.playerId || p.playerId == excludeId) continue;
                if (state.mode == MatchMode.Teams && attacker.team != null && attacker.team == p.team) continue;
                return true;
            }
            return false;
        }

        private static string ValidateTarget(MatchState state, PlayerState attacker, PlayerState target)
        {
            if (target == null || !target.alive) return "invalid_target";
            if (attacker.playerId == target.playerId) return "cannot_target_self";
            if (state.mode == MatchMode.Teams && attacker.team != null && attacker.team == target.team) return "friendly_fire_blocked";
            if (attacker.lastTargetedPlayerId == target.playerId && HasAlternativeTarget(state, attacker, target.playerId))
                return "repeat_target_blocked";
            return null;
        }

        public struct UseAttackResult
        {
            public bool ok;
            public string error;
            public CombatResolver.AttackOutcome outcome;
            public MatchResult result;
        }

        public static UseAttackResult UseAttack(MatchState state, int playerId, string rewardId, int targetId)
        {
            if (!state.players.TryGetValue(playerId, out var attacker) || !attacker.alive)
                return new UseAttackResult { ok = false, error = "not_in_match" };
            state.players.TryGetValue(targetId, out var target);
            var targetError = ValidateTarget(state, attacker, target);
            if (targetError != null) return new UseAttackResult { ok = false, error = targetError };
            if (attacker.pendingReward == null || attacker.pendingReward.rewardId != rewardId || attacker.pendingReward.type != RewardType.AttackChoice)
                return new UseAttackResult { ok = false, error = "no_such_reward" };

            var consumed = RewardResolver.ConsumeReward(attacker, rewardId);
            if (!consumed.ok) return new UseAttackResult { ok = false, error = consumed.error };
            attacker.lastTargetedPlayerId = targetId;

            var ability = CombatResolver.AttackFor(attacker.characterId);
            var outcome = CombatResolver.ResolveAttack(attacker, target, ability);

            var result = CheckWinCondition(state);
            state.result = result;
            if (result != null) state.status = MatchStatus.Completed;

            return new UseAttackResult { ok = true, outcome = outcome, result = result };
        }

        public struct UseFreezeResult
        {
            public bool ok;
            public string error;
        }

        /// Freezes the target: their next correct answer won't advance them. Doesn't
        /// touch HP/position itself, so — unlike attack/bonus_move — it can never itself
        /// produce a match result.
        public static UseFreezeResult UseFreeze(MatchState state, int playerId, string rewardId, int targetId)
        {
            if (!state.players.TryGetValue(playerId, out var attacker) || !attacker.alive)
                return new UseFreezeResult { ok = false, error = "not_in_match" };
            state.players.TryGetValue(targetId, out var target);
            var targetError = ValidateTarget(state, attacker, target);
            if (targetError != null) return new UseFreezeResult { ok = false, error = targetError };
            if (attacker.pendingReward == null || attacker.pendingReward.rewardId != rewardId || attacker.pendingReward.type != RewardType.Freeze)
                return new UseFreezeResult { ok = false, error = "no_such_reward" };

            var consumed = RewardResolver.ConsumeReward(attacker, rewardId);
            if (!consumed.ok) return new UseFreezeResult { ok = false, error = consumed.error };
            attacker.lastTargetedPlayerId = targetId;

            target.frozen = true;
            return new UseFreezeResult { ok = true };
        }

        public struct WaiveRewardResult
        {
            public bool ok;
            public string error;
        }

        /// Clears a pending reward with no other side effect — used when there's no
        /// valid target to act on it (e.g. attack_choice/freeze with no living
        /// opponent left). Mirrors server/src/matchEngine/MatchEngine.ts's
        /// waiveReward — without this, an un-consumed reward stays pending and blocks
        /// every future reward roll until it naturally expires.
        public static WaiveRewardResult WaiveReward(MatchState state, int playerId, string rewardId)
        {
            if (!state.players.TryGetValue(playerId, out var player))
                return new WaiveRewardResult { ok = false, error = "not_in_match" };
            if (player.pendingReward == null || player.pendingReward.rewardId != rewardId)
                return new WaiveRewardResult { ok = false, error = "no_such_reward" };

            var consumed = RewardResolver.ConsumeReward(player, rewardId);
            return new WaiveRewardResult { ok = consumed.ok, error = consumed.error };
        }

        public struct ConsumeBonusMoveResult
        {
            public bool ok;
            public string error;
            public GridPos newPos;
            public bool goalReached;
            public MatchResult result;
        }

        /// Consuming a bonus_move reward advances the player extra steps up their lane
        /// immediately — there's no separate directional "move" action anymore since a
        /// lane only has one direction to go.
        public static ConsumeBonusMoveResult ConsumeBonusMove(MatchState state, int playerId, string rewardId)
        {
            if (!state.players.TryGetValue(playerId, out var player) || !player.alive)
                return new ConsumeBonusMoveResult { ok = false, error = "not_in_match" };
            if (player.pendingReward == null || player.pendingReward.rewardId != rewardId || player.pendingReward.type != RewardType.BonusMove)
                return new ConsumeBonusMoveResult { ok = false, error = "no_such_reward" };

            var consumed = RewardResolver.ConsumeReward(player, rewardId);
            if (!consumed.ok) return new ConsumeBonusMoveResult { ok = false, error = consumed.error };

            var character = CharacterCatalog.Get(player.characterId);
            int bonus = character.abilityType == QuizBattle.Characters.AbilityType.Passive && character.bonusMoveSteps > 0
                ? character.bonusMoveSteps
                : MatchState.DefaultBonusMoveSteps;
            AdvancePlayer(player, state.gridHeight, bonus);

            var result = CheckWinCondition(state);
            state.result = result;
            if (result != null) state.status = MatchStatus.Completed;

            return new ConsumeBonusMoveResult { ok = true, newPos = player.pos, goalReached = player.goalReached, result = result };
        }

        private static Dictionary<object, List<PlayerState>> LivingGroups(MatchState state)
        {
            var groups = new Dictionary<object, List<PlayerState>>();
            foreach (var player in state.players.Values)
            {
                if (!player.alive) continue;
                object key = state.mode == MatchMode.Teams && player.team != null ? player.team : player.playerId;
                if (!groups.TryGetValue(key, out var list))
                {
                    list = new List<PlayerState>();
                    groups[key] = list;
                }
                list.Add(player);
            }
            return groups;
        }

        public static MatchResult CheckWinCondition(MatchState state)
        {
            var groups = LivingGroups(state);

            // Race win takes priority: first living player (or, in teams mode, any
            // member of a team) to reach the goal row wins outright.
            foreach (var (key, players) in groups)
            {
                if (players.Any(p => p.goalReached))
                {
                    return new MatchResult { winnerId = key, reason = WinReason.Goal };
                }
            }

            if (groups.Count == 1)
            {
                return new MatchResult { winnerId = groups.Keys.First(), reason = WinReason.Hp };
            }
            if (groups.Count == 0)
            {
                return new MatchResult { winnerId = null, reason = WinReason.Hp };
            }

            // Forced decision once any living player has been asked enough questions —
            // replaces the old round-cap zone-control tiebreak with lane progress
            // (furthest up the board wins, tiebroken by total HP).
            bool anyoneAtCap = state.players.Values.Any(p => p.alive && p.questionsAnswered >= state.maxRounds);
            if (anyoneAtCap)
            {
                object bestKey = null;
                int bestProgress = -1;
                int bestHp = -1;
                foreach (var (key, players) in groups)
                {
                    int progress = players.Max(p => p.pos.y);
                    int hp = players.Sum(p => p.hp);
                    if (progress > bestProgress || (progress == bestProgress && hp > bestHp))
                    {
                        bestKey = key;
                        bestProgress = progress;
                        bestHp = hp;
                    }
                }
                return new MatchResult { winnerId = bestKey, reason = WinReason.Progress };
            }

            return null;
        }
    }
}

File: server/src/matchEngine/MatchEngine.ts
import {
  DEFAULT_BONUS_MOVE_STEPS,
  createMatchState,
  type GridPos,
  type MatchState,
  type PlayerState,
} from "./MatchState";
import { getCharacter } from "./characters/characterConfig";
import { rollReward, expireStaleRewards, consumeReward } from "./RewardResolver";
import { attackFor, resolveAttack, tickDot, type AttackOutcome } from "./CombatResolver";

export function createMatch(
  matchId: number,
  mode: "ffa" | "teams",
  maxRounds = 20,
  gridHeight?: number,
  timerSeconds = 30
): MatchState {
  return createMatchState(matchId, mode, maxRounds, gridHeight, timerSeconds);
}

export function addPlayer(
  state: MatchState,
  playerId: number,
  name: string,
  characterId: string,
  team: string | null,
  startPos: GridPos
): PlayerState {
  const character = getCharacter(characterId);
  const player: PlayerState = {
    playerId,
    name,
    characterId,
    team,
    hp: character.baseStats.hp,
    maxHp: character.baseStats.hp,
    pos: startPos,
    alive: true,
    consecutiveCorrect: 0,
    lastRewardType: null,
    pendingReward: null,
    pendingDot: null,
    totalCorrectAnswers: 0,
    maxStreak: 0,
    questionsAnswered: 0,
    currentQuestion: null,
    goalReached: false,
    finishRank: null,
    finishedAt: null,
    frozen: false,
    lastTargetedPlayerId: null,
  };
  state.players.set(playerId, player);
  return player;
}

export function startMatch(state: MatchState): void {
  state.status = "active";
}

/** Moves a player toward the goal row (grid.height - 1), clamped there, and
 * flags goalReached once they hit it. In 3+ player matches, records their finish rank. */
export function advancePlayer(player: PlayerState, state: MatchState, steps: number): void {
  const goalY = state.grid.height - 1;
  const nextY = Math.min(goalY, player.pos.y + steps);
  player.pos = { x: player.pos.x, y: nextY };
  if (nextY >= goalY && !player.goalReached) {
    player.goalReached = true;
    player.finishedAt = Date.now();
    if (!state.finishOrder.includes(player.playerId)) {
      state.finishOrder.push(player.playerId);
      player.finishRank = state.finishOrder.length;
    }
  }
}

export function pushQuestion(state: MatchState, playerId: number, questionId: number, correctIndex: number): void {
  const player = state.players.get(playerId);
  if (!player) return;
  player.currentQuestion = { questionId, correctIndex, questionNumber: player.questionsAnswered + 1 };
}

export interface SubmitAnswerResult {
  ok: boolean;
  error?: string;
  correct?: boolean;
  streakCount?: number;
  rewardOffered?: { rewardId: string; type: string } | null;
  newPos?: GridPos;
  goalReached?: boolean;
  dotDamage?: number;
  result?: MatchState["result"];
}

export function submitAnswer(
  state: MatchState,
  playerId: number,
  choiceIndex: number,
  rng: () => number = Math.random
): SubmitAnswerResult {
  const player = state.players.get(playerId);
  if (!player || !player.alive) return { ok: false, error: "not_in_match" };
  if (!player.currentQuestion) return { ok: false, error: "no_active_question" };

  const correct = choiceIndex === player.currentQuestion.correctIndex;
  player.currentQuestion = null;
  player.questionsAnswered += 1;

  player.consecutiveCorrect = correct ? player.consecutiveCorrect + 1 : 0;
  if (correct) {
    player.totalCorrectAnswers += 1;
    player.maxStreak = Math.max(player.maxStreak, player.consecutiveCorrect);
  }

  let rewardOffered: SubmitAnswerResult["rewardOffered"] = null;
  if (correct && player.consecutiveCorrect >= 2 && !player.pendingReward) {
    const reward = rollReward(player, player.questionsAnswered, rng);
    player.pendingReward = reward;
    rewardOffered = { rewardId: reward.rewardId, type: reward.type };
  }
  expireStaleRewards([player], player.questionsAnswered);

  // A correct answer is what drives you toward the goal — this replaces the
  // old movement-budget/direction system entirely. A frozen player still
  // answers/builds streak normally; only this one advance is consumed/skipped.
  if (correct) {
    if (player.frozen) {
      player.frozen = false;
    } else {
      advancePlayer(player, state, 1);
    }
  }

  const dotDamage = tickDot(player);

  const result = checkWinCondition(state);
  state.result = result;
  if (result) state.status = "completed";

  return {
    ok: true,
    correct,
    streakCount: player.consecutiveCorrect,
    rewardOffered,
    newPos: player.pos,
    goalReached: player.goalReached,
    dotDamage,
    result,
  };
}

/** A player's personal question timer expired with no answer submitted —
 * scored the same as a wrong answer (streak reset, no movement), but still
 * ticks their own DoT and advances their personal question count so the
 * match-length cap and reward-expiry clock keep moving even if they stall. */
export function timeoutAnswer(state: MatchState, playerId: number): SubmitAnswerResult {
  const player = state.players.get(playerId);
  if (!player || !player.alive || !player.currentQuestion) return { ok: false, error: "no_active_question" };

  player.currentQuestion = null;
  player.questionsAnswered += 1;
  player.consecutiveCorrect = 0;

  // A reward the player never acted on (missed/ignored the popup) must not linger —
  // submitAnswer's `!player.pendingReward` gate would otherwise silently block every
  // future streak reward until natural expiry, several questions later.
  if (player.pendingReward) {
    consumeReward(player, player.pendingReward.rewardId);
  }
  expireStaleRewards([player], player.questionsAnswered);

  const dotDamage = tickDot(player);

  const result = checkWinCondition(state);
  state.result = result;
  if (result) state.status = "completed";

  return {
    ok: true,
    correct: false,
    streakCount: 0,
    rewardOffered: null,
    newPos: player.pos,
    goalReached: player.goalReached,
    dotDamage,
    result,
  };
}

/** Every living player attacker could legally target (alive, not self, not a
 * teammate, and NOT already finished) other than excludeId — used both to enforce
 * "can't target the same player twice in a row" and to waive that rule when there's
 * genuinely no one else to pick, rather than soft-locking the reward entirely. */
function hasAlternativeTarget(state: MatchState, attacker: PlayerState, excludeId: number): boolean {
  for (const p of state.players.values()) {
    if (!p.alive || p.goalReached || p.playerId === attacker.playerId || p.playerId === excludeId) continue;
    if (state.mode === "teams" && attacker.team && attacker.team === p.team) continue;
    return true;
  }
  return false;
}

function validateTarget(state: MatchState, attacker: PlayerState, target: PlayerState | undefined): string | null {
  if (attacker.goalReached) return "attacker_already_finished";
  if (!target || !target.alive) return "invalid_target";
  if (target.goalReached) return "target_already_finished";
  if (attacker.playerId === target.playerId) return "cannot_target_self";
  if (state.mode === "teams" && attacker.team && attacker.team === target.team) return "friendly_fire_blocked";
  if (attacker.lastTargetedPlayerId === target.playerId && hasAlternativeTarget(state, attacker, target.playerId)) {
    return "repeat_target_blocked";
  }
  return null;
}

export interface UseAttackResult {
  ok: boolean;
  error?: string;
  outcome?: AttackOutcome;
  result?: MatchState["result"];
}

export function useAttack(state: MatchState, playerId: number, rewardId: string, targetId: number): UseAttackResult {
  const attacker = state.players.get(playerId);
  const target = state.players.get(targetId);
  if (!attacker || !attacker.alive) return { ok: false, error: "not_in_match" };
  const targetError = validateTarget(state, attacker, target);
  if (targetError) return { ok: false, error: targetError };
  if (!attacker.pendingReward || attacker.pendingReward.rewardId !== rewardId || attacker.pendingReward.type !== "attack_choice") {
    return { ok: false, error: "no_such_reward" };
  }

  const consumed = consumeReward(attacker, rewardId);
  if (!consumed.ok) return { ok: false, error: consumed.error };
  attacker.lastTargetedPlayerId = targetId;

  const ability = attackFor(attacker.characterId);
  const outcome = resolveAttack(attacker, target!, ability);

  const result = checkWinCondition(state);
  state.result = result;
  if (result) state.status = "completed";

  return { ok: true, outcome, result };
}

export interface UseFreezeResult {
  ok: boolean;
  error?: string;
}

/** Freezes the target: their next correct answer won't advance them. Doesn't
 * touch HP/position itself, so — unlike attack/bonus_move — it can never
 * itself produce a match result. */
export function useFreeze(state: MatchState, playerId: number, rewardId: string, targetId: number): UseFreezeResult {
  const attacker = state.players.get(playerId);
  const target = state.players.get(targetId);
  if (!attacker || !attacker.alive) return { ok: false, error: "not_in_match" };
  const targetError = validateTarget(state, attacker, target);
  if (targetError) return { ok: false, error: targetError };
  if (!attacker.pendingReward || attacker.pendingReward.rewardId !== rewardId || attacker.pendingReward.type !== "freeze") {
    return { ok: false, error: "no_such_reward" };
  }

  const consumed = consumeReward(attacker, rewardId);
  if (!consumed.ok) return { ok: false, error: consumed.error };
  attacker.lastTargetedPlayerId = targetId;

  target!.frozen = true;
  return { ok: true };
}

export interface ConsumeBonusMoveResult {
  ok: boolean;
  error?: string;
  newPos?: GridPos;
  goalReached?: boolean;
  result?: MatchState["result"];
}

/** Consuming a bonus_move reward advances the player extra steps up their
 * lane immediately — there's no separate directional "move" action anymore
 * since a lane only has one direction to go. */
export function consumeBonusMove(state: MatchState, playerId: number, rewardId: string): ConsumeBonusMoveResult {
  const player = state.players.get(playerId);
  if (!player || !player.alive) return { ok: false, error: "not_in_match" };
  if (!player.pendingReward || player.pendingReward.rewardId !== rewardId || player.pendingReward.type !== "bonus_move") {
    return { ok: false, error: "no_such_reward" };
  }
  const consumed = consumeReward(player, rewardId);
  if (!consumed.ok) return { ok: false, error: consumed.error };

  const character = getCharacter(player.characterId);
  const bonus = character.ability.type === "passive" ? character.ability.bonusMoveSteps ?? DEFAULT_BONUS_MOVE_STEPS : DEFAULT_BONUS_MOVE_STEPS;
  advancePlayer(player, state, bonus);

  const result = checkWinCondition(state);
  state.result = result;
  if (result) state.status = "completed";

  return { ok: true, newPos: player.pos, goalReached: player.goalReached, result };
}

export interface WaiveRewardResult {
  ok: boolean;
  error?: string;
}

/** Clears a pending reward with no other side effect — used when the client can't
 * act on it (e.g. an attack_choice/freeze offered with no living opponent left to
 * target). Without this, an un-consumable reward stays pending and blocks every
 * future reward roll (see submitAnswer's `!player.pendingReward` gate) until it
 * naturally expires several questions later. */
export function waiveReward(state: MatchState, playerId: number, rewardId: string): WaiveRewardResult {
  const player = state.players.get(playerId);
  if (!player) return { ok: false, error: "not_in_match" };
  if (!player.pendingReward || player.pendingReward.rewardId !== rewardId) {
    return { ok: false, error: "no_such_reward" };
  }
  return consumeReward(player, rewardId);
}

function livingGroups(state: MatchState): Map<string | number, PlayerState[]> {
  const groups = new Map<string | number, PlayerState[]>();
  for (const player of state.players.values()) {
    if (!player.alive) continue;
    const key = state.mode === "teams" && player.team ? player.team : player.playerId;
    const list = groups.get(key) ?? [];
    list.push(player);
    groups.set(key, list);
  }
  return groups;
}

export function checkWinCondition(state: MatchState): MatchState["result"] {
  const groups = livingGroups(state);

  // 1. All players eliminated (draw / complete wipeout)
  if (groups.size === 0) {
    return { winnerId: null, reason: "hp" };
  }

  // Active racers: living players who have NOT yet reached the goal
  const activeLivingPlayers = [...state.players.values()].filter((p) => p.alive && !p.goalReached);

  // Group active racers by team (or playerId in FFA)
  const activeGroups = new Map<string | number, PlayerState[]>();
  for (const p of activeLivingPlayers) {
    const key = state.mode === "teams" && p.team ? p.team : p.playerId;
    const list = activeGroups.get(key) ?? [];
    list.push(p);
    activeGroups.set(key, list);
  }

  // 2. End when there is only one player/team left actively racing
  if (state.players.size > 1) {
    if (activeGroups.size <= 1) {
      // If someone reached the goal, record any remaining active runner-up and finish
      if (state.finishOrder.length > 0) {
        for (const p of activeLivingPlayers) {
          if (!state.finishOrder.includes(p.playerId)) {
            state.finishOrder.push(p.playerId);
            p.finishRank = state.finishOrder.length;
          }
        }
        const winnerId = state.mode === "teams"
          ? (state.players.get(state.finishOrder[0])?.team ?? state.finishOrder[0])
          : state.finishOrder[0];
        return { winnerId, reason: "goal" };
      }

      // No one reached the goal, but only 1 living player/team remains (elimination by HP)
      if (groups.size === 1) {
        const [winnerId] = groups.keys();
        return { winnerId, reason: "hp" };
      }
    }
  } else {
    // Single player match: ends when the player reaches the goal or is eliminated
    if (activeGroups.size === 0) {
      if (state.finishOrder.length > 0) {
        return { winnerId: state.finishOrder[0], reason: "goal" };
      }
      return { winnerId: null, reason: "hp" };
    }
  }

  // Forced decision once any living player has been asked enough questions
  const anyoneAtCap = [...state.players.values()].some((p) => p.alive && p.questionsAnswered >= state.maxRounds);
  if (anyoneAtCap) {
    let bestKey: string | number | null = null;
    let bestProgress = -1;
    let bestHp = -1;
    for (const [key, players] of groups) {
      const progress = Math.max(...players.map((p) => p.pos.y));
      const hp = players.reduce((sum, p) => sum + p.hp, 0);
      if (progress > bestProgress || (progress === bestProgress && hp > bestHp)) {
        bestKey = key;
        bestProgress = progress;
        bestHp = hp;
      }
    }
    return { winnerId: bestKey, reason: "progress" };
  }

  return null;
}
